src/main/java/com/urbanairship/datacube/dbharnesses/HbaseBatchIncrementer.java

```diff
@@ -0,0 +1,248 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Package private class implementing all the batch increment operations accomplished by the hbase dbharness
+ *
+ * @param <T>
+ */
+class HbaseBatchIncrementer<T extends Op> {
+
+    private final HTablePool pool;
+    private final IncrementerMetrics incrementerMetrics;
+    private final BlockingIO<Address, BoxedByteArray> serializer;
+    private HbaseDbHarnessConfiguration configuration;
+
+    interface BlockingIO<I, O> {
+        O apply(I i) throws InterruptedException, IOException;
+    }
+
+    HbaseBatchIncrementer(HbaseDbHarnessConfiguration configuration, HTablePool pool, BlockingIO<Address, BoxedByteArray> serializer) {
+        this.configuration = configuration;
+        this.pool = pool;
+        this.serializer = serializer;
+        this.incrementerMetrics = new IncrementerMetrics(HBaseDbHarness.class, configuration.metricsScope);
+    }
+
+    /**
+     * Slices up the batch map into batches of a reasonable size to send to the database, interprets the results
+     * (include partial failures where only a part of the batch succeeded, and an exception was thrown) and updates
+     * the success map and list with the results. It then rethrows whatever exception resulted from the HBase
+     * interaction.
+     *
+     * @param batchMap            The write operations to accomplish
+     * @param successfulAddresses The datacube address whose writes succeeded
+     * @param successfulRows      A map from HBase row key to the value in the row after completion of the increment
+     *                            operation
+     *
+     * @throws IOException
+     * @throws InterruptedException
+     */
+
+    public void batchIncrement(Map<Address, T> batchMap, List<Address> successfulAddresses, Map<byte[], byte[]> successfulRows) throws IOException, InterruptedException {
+        Iterable<List<Map.Entry<Address, T>>> partitions = Iterables.partition(batchMap.entrySet(), configuration.batchSize);
+
+        CubeIncrementResults.Builder cubeIncrementResultsBuilder = CubeIncrementResults.newBuilder();
+
+        int batchesPerFlush = 0;
+        Map<BoxedByteArray, Address> backwards = new HashMap<>();
+        for (List<Map.Entry<Address, T>> entries : partitions) {
+            Map<BoxedByteArray, T> increments = new HashMap<>();
+            batchesPerFlush++;
+            final long nanoTimeBeforeWrite = System.nanoTime();
+            for (Map.Entry<Address, T> entry : entries) {
+                final Address address = entry.getKey();
+                final BoxedByteArray rowKey = serializer.apply(address);
+
+                increments.put(rowKey, entry.getValue());
+                backwards.put(rowKey, address);
+            }
+
+            cubeIncrementResultsBuilder.add(this.hbaseIncrement(increments));
```

Comment:
As written this will blow away all existing state in the builder, unless I'm mistaken.

Reply:
Correct, but you shouldn't do that

Reply:
As in adding a results object to a builder that already has exceptions set isn't intended behavior. I can make it throw?